nomt/src/io/page_pool.rs

```diff
@@ -158,44 +189,63 @@ impl PagePool {
         self.inner.freelist.write().push(page);
     }
 
-    fn grow(&self, freelist: &mut Vec<Page>) -> Page {
-        assert!(freelist.is_empty());
+    fn grow(&self, freelist_guard: &mut RwLockWriteGuard<Vec<Page>>) -> Page {
+        assert!(freelist_guard.is_empty());
+
+        // First step is to allocate a new region.
         let region_ptr = unsafe {
             libc::mmap(
                 std::ptr::null_mut(),
-                REGION_SIZE,
+                REGION_BYTE_SIZE,
                 libc::PROT_READ | libc::PROT_WRITE,
                 libc::MAP_PRIVATE | libc::MAP_ANONYMOUS,
                 /* fd */ -1,
                 /* offset */ 0,
             )
         };
+        assert!(!region_ptr.is_null());
         if region_ptr == libc::MAP_FAILED {
             panic!("Failed to allocate memory");
         }
-        let mut regions = self.inner.regions.write();
-        regions.push(region_ptr);
 
-        let region: u32 = ((regions.len() - 1) as u32) << 16;
-        for slot in 0..=REGION_SLOT_MASK {
-            freelist.push(Page(PageIndex(slot | region)));
+        // Next, we need to store the region pointer in the regions array.
+        //
+        // We store the pointer in the regions array before incrementing n_regions. This is not
+        // strictly necessary, because the freelist is empty and no page can refer to the new region
+        // yet. Likewise, drop cannot happen during this operation. We still do it in this order
+        // to just err on the safe side and avoid any potential issues.
+        let region_ix = self.inner.n_regions.load(Ordering::Relaxed);
+        self.inner.regions[region_ix as usize].store(region_ptr as *mut u8, Ordering::Release);
+        self.inner.n_regions.fetch_add(1, Ordering::Relaxed);
+
+        // Finally, we need to populate the freelist with the pages in the new region.
+        for slot in 0..SLOTS_PER_REGION {
+            freelist_guard.push(Page(PageIndex::from_region_and_slot(
+                region_ix,
+                slot as u32,
+            )));
         }
-        // unwrap: we know that the freelist is not empty
-        freelist.pop().unwrap()
+        // UNWRAP: we know that the freelist is not empty, because we just filled it.
+        freelist_guard.pop().unwrap()
     }
 
     fn data_ptr(&self, page_index: PageIndex) -> *mut u8 {
-        let region_ptr = self.inner.regions.read()[page_index.region()];
-        unsafe { region_ptr.add(page_index.slot_index() * PAGE_SIZE) as *mut u8 }
+        let region = page_index.region();
+        assert!(region < self.inner.n_regions.load(Ordering::Relaxed) as usize);
```

Comment:
Just like the other comment. What we want here is to load this with `Acquire` ordering and then load the region pointer with `Relaxed`.

Reply:
I suspect this is false. Orderings essentially specify how the memory outside of the atomic in question "synchronized", but the atomic itself is always "synchronized". That being said, I also feel that putting things in proper release-acquire brackets feels more right.

Reply:
No, it's not false, when you account for compiler and CPU reordering. Proper acquire/release brackets are really important.

Atomicity guarantees operation ordering, nothing else. So in the first iteration that I commented on, it would be possible to read `n_regions=1` but `regions[1].load()` could be null. This could happen if the compiler or CPU reordered the stores, such that `n_regions` was incremented prior to storing the pointer in `regions[1]`, which was permitted by the `Release`.

Your mental model should be roughly that writes can move up across a release but not down below it. And that reads can move down across an acquire but not up above it.